_algorithm_design/on_computing_next_greater_intervals_in_linear_time.md

```diff
@@ -0,0 +1,179 @@
+
+Let's say we have separate lists `leftBound` and `rightBound` such that the $leftBound[i]$ is the index of the left bound of $list[i]$ and $rightBound[i]$ is the index of the right bound respectively. If an element $list[i]$ doesn't have a sound left bound then we set $leftBound[i]$ to $-1$; similarly we set $rightBound[i]$ to $N$.
+
+## Motivation for the problem
+
+Many programming problems can be reduced to the problem statement that we defined. Hence the solution to the above can be used as a sub-routine in the solution of the original problem. I'll illustrate a well-known problem to motivate the development of a solution to our problem statement.
+
+Check out [this][spoj-histogra] problem from spoj. After a few preliminary sanity checks, you can convince yourself that every entry in the histogram forms a rectangle in its `nextLesserInterval`. The `nextLesserInterval` is just `nextGreaterInterval` with respect to the $<$ operator. Hence each such rectangle will be a candidate for max-area rectangle.
+
+## Designing the machinery
+
+It typically helps to solve a simpler version of the original problem to get started. More complexity can be added step-by-step to reach the solution for the original problem. Hence, I first consider the problem for computing $leftBound$ separately.
+
+### Computing the leftBound list
+
+A naive way to compute left bound for an element $list[i]$ is to linearly search leftward in $list$ till you find the first element greater than $list[i]$. In the worst case there are $N$ states to check for every element in $list$, hence a total of $O(N^{2})$ states to search. Thus it'll take $O(N^{2})$ time to compute the list $leftBound$.
+
+However, say we already have $leftBound$ computed for the entire $list$ but we get a new element `curr` to be appended to $list$. There can be two cases here:
+- case1: $curr < list[N-1]$
+    - In this case $list[N-1]$ will trivially be the left bound of $curr$.
+- case2: $curr >= list[N-1]$
+    - Here we can use the naive way of searching linearly leftwards in $list$ till we find the first element greater than $curr$. However, one can notice from the definition of $leftBound$ that the elements in $list$ between the indices $N-1$ and $leftBound[N-1]$ are all smaller than $curr$.
+    - Thus instead of searching linearly leftwards, we could skip the elements in between and directly check $list[leftBound[N-1]]$, then $list[leftBound[leftBound[N-1]]]$ and so on until we find the element greater than $curr$.
+    - The set of indices $list[N-1]$, $list[leftBound[N-1]]$, $list[leftBound[leftBound[N-1]]]$ and so on form a topology of ancestors dictated strictly by ordering with respect to the $>$ operator. At this point we can safely conclude that computation of left bound for a new element depends strictly on the topological ancestors of the last element in the $list$.
+
+Hence we have pruned the search space but by how much? We will discuss the answer to this shortly.
```

Comment:
Hmm... that was the original idea. However I wanted it to look like an inductive step which helps us write the algorithm instead of an intermediate step of a known algorithm.

I'll see if I can reframe this part.

Reply:
If you want it to feel like the inductive step, try to use some other variable and not `N` because that is already used in another context (Size of the list here.)

Reply:
Sure. Changing it.